core/sqf/src/seatrans/hbase-trx/src/main/java/org/apache/hadoop/hbase/coprocessor/transactional/TrxRegionEndpoint.java.tmpl

```diff
@@ -7335,6 +7338,245 @@ TrxTransactionState isTransactionOnCommittedList(final long transactionId)
+  // to estimate how many entries make up a row on average.
+
+  @Override
+  public void trafEstimateRowCount(RpcController controller,
+			           TrafEstimateRowCountRequest request,
+			           RpcCallback<TrafEstimateRowCountResponse> done) {
+
+    TrafEstimateRowCountResponse response = null;
+    Throwable t = null;
+
+    int numCols = request.getNumCols();
+
+    // To estimate incidence of nulls, read the first 500 rows worth
+    // of KeyValues.
+    final int ROWS_TO_SAMPLE = 500;
+    int putKVsSampled = 0;
+    int nonPutKVsSampled = 0;
+    int missingKVsCount = 0;
+    int sampleRowCount = 0;
+    long totalEntries = 0;   // KeyValues in all HFiles for table
+    long totalSizeBytes = 0; // Size of all HFiles for table 
+    long estimatedTotalPuts = 0;
+    boolean more = true;
+    long estimatedRowCount = 0;
+
+    // Access the file system to go directly to the table's HFiles.
+    // Create a reader for the file to access the KV entry count and
+    // size in bytes stored in the trailer block.
+    
+    // For aligned format tables, the number of rows equals the
+    // number of KeyValue entries. For non-aligned format, it's
+    // more complicated. There is a KeyValue entry for each 
+    // column value, except the KeyValue may be missing because
+    // the column has a null value or because the column has a
+    // default value that has not been materialized.
+
+    // For non-aligned format tables, we sample some rows and
+    // count how many entries there are per row, so our caller
+    // can estimate the average number of missing values per row.
+    // Once our caller has that estimate, it can estimate the
+    // number of rows.
+
+    // We only do the sampling for non-aligned tables (numCols > 1),
+    // and we only do it on the first HFile of the first Region.
+    // The first Region is detected by having a null start key.
+
+    CacheConfig cacheConf = new CacheConfig(config);
+    byte[] startKey = regionInfo.getStartKey();
+
+    // Get the list of store files in column family '#1'. There might
+    // not be any. For example, a new Trafodion table might be entirely
+    // in memstore with nothing written out yet. Or we may be accessing
+    // a native HBase table which lacks the '#1' colum family.
+    List<String> storeFileList = null;
+    try {
+      byte[] familyName = "#1".getBytes();
+      byte[][] familyNames = { familyName };
+      storeFileList = m_Region.getStoreFileList(familyNames);
+    }
+    catch (IllegalArgumentException iae) {
+      // this gets thrown when the column family doesn't exist;
+      // we'll just use an empty list instead
+      storeFileList = new ArrayList<String>();
+    }
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Trafodion estimate row count sees " + storeFileList.size() + " files.");
+      for (String sfn : storeFileList) {
+        LOG.debug("*** " + sfn);
+      }
+      if (startKey == null)
```

Comment:
That's what I found as well during unit test. I saw a non-null startkey with a length of zero. Since this is debugging code, though, I wrote it to take into account both possibilities. (In fact, this debug code is how I found out that you are right!)